Add TensorRT engine to ONNX Runtime GPU documentation (#657)
huggingface/optimum: docs/source/onnxruntime/usage_guides/gpu.mdx

something is wrong with the TensorRT or ONNX Runtime installation.

### TensorRT engine build and warmup

TensorRT requires to build its [inference engine](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#build-phase) ahead of inference, which takes some time due to the model optimization, nodes fusion. To avoid rebuiling the engine every time the model is loaded, ONNX Runtime provides a pair of options to allow to save the engine: `trt_engine_cache_enable` and `trt_engine_cache_path`.

> We recommend to set these two provier options when using TensorRT execution provider. The usage is as follow, where here [`optimum/gpt2`](https://huggingface.co/optimum/gpt2) is an ONNX model converted from PyTorch using [Optimum ONNX export](https://huggingface.co/docs/optimum/main/en/exporters/onnx/usage_guides/export_a_model):

Reviewer: ```suggestion
We recommend setting these two provider options when using the TensorRT execution provider. The usage is as follows, where [`optimum/gpt2`](https://huggingface.co/optimum/gpt2) is an ONNX model converted from PyTorch using the [Optimum ONNX exporter](https://huggingface.co/docs/optimum/main/en/exporters/onnx/usage_guides/export_a_model):
```

```python
from optimum.onnxruntime import ORTModelForCausalLM

provider_options = {
    "trt_engine_cache_enable": True,
    "trt_engine_cache_path": "/path/to/trt_cache_gpt2_example"
}

# the TensorRT engine is not built here, it will be when doing inference
ort_model = ORTModelForCausalLM.from_pretrained(
    "optimum/gpt2",
    use_cache=False,
    provider="TensorrtExecutionProvider",
    provider_options=provider_options
)
```

> TensorRT builds its engine depending on specified input shapes. Unfortunately, in [current ONNX Runtime implementation](https://github.com/microsoft/onnxruntime/blob/613920d6c5f53a8e5e647c5f1dcdecb0a8beef31/onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc#L1677-L1688) (references: [1](https://github.com/microsoft/onnxruntime/issues/13559), [2](https://github.com/microsoft/onnxruntime/issues/13851)), the engine is rebuilt every time an input has a shape smaller than the previously smallest encountered shape, and conversely if the input has a shape larger than the previously largest encountered shape. For example, if a model takes `(batch_size, input_ids)` as inputs, and the model takes successively the inputs:

Reviewer: ```suggestion
TensorRT builds its engine depending on specified input shapes. Unfortunately, in the [current ONNX Runtime implementation](https://github.com/microsoft/onnxruntime/blob/613920d6c5f53a8e5e647c5f1dcdecb0a8beef31/onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc#L1677-L1688) (references: [1](https://github.com/microsoft/onnxruntime/issues/13559), [2](https://github.com/microsoft/onnxruntime/issues/13851)), the engine is rebuilt every time an input has a shape smaller than the previously smallest encountered shape, and conversely if the input has a shape larger than the previously largest encountered shape. For example, if a model takes `(batch_size, input_ids)` as inputs, and the model takes successively the inputs:
```

> 1. `input.shape: (4, 5)  --> the engine is built (first input)`
> 2. `input.shape: (4, 10) --> engine rebuilt (8 larger than 3)`
> 3. `input.shape: (4, 7)  --> no rebuild (3 <= 5 <= 8)`
> 4. `input.shape: (4, 12) --> engine rebuilt (9 <= 10)`
> 5. `input.shape: (4, 3)  --> engine rebuilt (3 <= 5)`

Reviewer: Maybe I didn't understand the example, but shoudn't it be the following? I think there is a mismatch between the given input shapes and the values used in the explanations.
```suggestion
1. `input.shape: (4, 5)  --> the engine is built (first input)`
2. `input.shape: (4, 10) --> engine rebuilt (10 larger than 5)`
3. `input.shape: (4, 7)  --> no rebuild (5 <= 7 <= 10)`
4. `input.shape: (4, 12) --> engine rebuilt (10 <= 12)`
5. `input.shape: (4, 3)  --> engine rebuilt (3 <= 5)`
```

Author: yes my bad, I changed the numbers midway!

> One big issue is that engine building can be time consuming, especially for large models. Therefore, as a workaround, one recommendation is to **first build the TensorRT engine with an input of small shape, and an input of large shape to have an engine valid for all shapes inbetween**. This allows to avoid rebuilding the engine for new small and large shapes, which is unwanted once the model is deployed for inference.

Reviewer: ```suggestion
One big issue is that building the engine can be time consuming, especially for large models. Therefore, as a workaround, one recommendation is to **first build the TensorRT engine with an input of small shape, and then with an input of large shape to have an engine valid for all shapes inbetween**. This allows to avoid rebuilding the engine for new small and large shapes, which is unwanted once the model is deployed for inference.
```

Passing the engine cache path in the provider options, the engine can therefore be built once for all and used fully for inference thereafter.

> For example, for text generation, the engine building can be done with:

Reviewer: ```suggestion
For example, for text generation, the engine can be built with:
```

```python
from transformers import AutoTokenizer
from optimum.onnxruntime import ORTModelForCausalLM

provider_options = {
    "trt_engine_cache_enable": True,
    "trt_engine_cache_path": "/path/to/trt_cache_gpt2_example"
}

ort_model = ORTModelForCausalLM.from_pretrained(
    "optimum/gpt2",
    use_cache=False,
    provider="TensorrtExecutionProvider",
    provider_options=provider_options,
)
tokenizer = AutoTokenizer.from_pretrained("optimum/gpt2")

# build engine for a short sequence
text = ["short"]
encoded_input = tokenizer(text, return_tensors="pt").to("cuda")
output = ort_model(**encoded_input)

# build engine for a long sequence
text = [" a very long input just for demo purpose, this is very long" * 10]
encoded_input = tokenizer(text, return_tensors="pt").to("cuda")
output = ort_model(**encoded_input)
```

The engine is stored as:

![TensorRT engine cache folder](https://huggingface.co/datasets/optimum/documentation-images/resolve/main/onnxruntime/tensorrt_cache.png)

> Once the engine is built, generation is fast and does not need engine rebuild:

Reviewer: ```suggestion
Once the engine is built, generation is fast and does not need to rebuild the engine:
```
